uxarray/plot/dataarray_plot.py

```diff
@@ -179,60 +282,146 @@ def _point_raster(uxda: UxDataArray,
+                                   colorbar=colorbar,
+                                   cmap=cmap,
+                                   **kwargs)
+
+    else:
+        raise ValueError(
+            f"Invalid backend selected. Expected one of ['matplotlib', 'bokeh'] but received {backend}."
+        )
+
+    return raster
+
+
+def polygons(uxda: UxDataArray,
+             backend: Optional[str] = "bokeh",
+             exclude_antimeridian: Optional[bool] = True,
+             width: Optional[int] = 1000,
+             height: Optional[int] = 500,
+             colorbar: Optional[bool] = True,
+             cmap: Optional[str] = "Blues",
+             **kwargs):
+    """Vector Polygon Plot of a Data Variable Residing on an Unstructured Grid.
 
     Parameters
     ----------
-    method: str
-        Selects what type of element to rasterize (point, trimesh, polygon), with "point" being the only currently
-        implemented method.
     backend: str
         Selects whether to use Holoview's "matplotlib" or "bokeh" backend for rendering plots
-    projection: ccrs
-         Custom projection to transform (lon, lat) coordinates for rendering
-    pixel_ratio: float
-        Determines the resolution of the outputted raster.
+    exclude_antimeridian: bool,
+        Whether to exclude faces that cross the antimeridian (Polygon Raster Only)
     height: int
         Plot Height for Bokeh Backend
     width: int
         Plot Width for Bokeh Backend
-    cache: bool
-            Determines where computed elements (i.e. points, polygons) should be cached internally for subsequent plotting
-            calls
+    """
+    if not exclude_antimeridian:
+        warnings.warn(
+            "Including Antimeridian Polygons may lead to visual artifacts. It is suggested to keep "
+            "'exclude_antimeridian' set to True.")
 
-    Notes
-    -----
-    For further information about supported keyword arguments, please refer to the [Holoviews Documentation](https://holoviews.org/_modules/holoviews/operation/datashader.html#rasterize)
-    or run holoviews.help(holoviews.operation.datashader.rasterize).
+    gdf = uxda.to_geodataframe(exclude_antimeridian=exclude_antimeridian)
+
+    hv_polygons = hv.Polygons(gdf, vdims=[uxda.name])
+
+    if backend == "matplotlib":
+        # use holoviews matplotlib backend
+        hv.extension("matplotlib")
+
+        return hv_polygons.opts(colorbar=colorbar, cmap=cmap, **kwargs)
+
+    elif backend == "bokeh":
+        # use holoviews bokeh backend
+        hv.extension("bokeh")
+        return hv_polygons.opts(width=width,
+                                height=height,
+                                colorbar=colorbar,
+                                cmap=cmap,
+                                **kwargs)
+
+
+def nodes(uxda: UxDataArray,
+          backend: Optional[str] = "bokeh",
+          width: Optional[int] = 1000,
+          height: Optional[int] = 500,
+          **kwargs):
```

Comment:
I'd expect to see this function just calling `uxda.uxgrid.plot.nodes()` instead of implementing things itself. 

Its current implementation seems like you're trying to plot not only the Grid nodes' coords but also their corresponding values. I think such a functionality should be implemented with something like `UxDataArray.plot.values()`, and `UxDataArray.plot.nodes()` should only focus on the Grid geometry, hence my expectation with calling `uxda.uxgrid.plot.nodes()`. For future, `UxDataArray.plot.values()` would show node coords and values, or face coords and values (still as a point-plot), depending on data being node-centered or facec-centered, respectively.

Reply:
`uxda.uxgrid.plot.nodes()` is equivalent to `uxgrid.plot.nodes()`, which carries no information about the data variable.

`UxDataArray.plot.values()` sounds like what about default `UxDataArray.plot()` method, where depending on what element our data is mapped to, it returns the appropriate plot.

Having `UxDataArray.plot.values()` return a node plot with face-centered data may be misleading, because the data isn't really mapped to the centroid, it's meant to shade it's corresponding face.

There's an expectation that _any_ plotting method directly called on a `Grid` instance will be data-less (i.e. only the geometry). Similarly, _any_ plot method called directly on a `UxDataArray` will have a data variable. 

### Plot Nodes without Data
```Python
uxds.uxgrid.plot.nodes()
```

### Plot Nodes with Data
```Python
uxds['d_var'].plot.nodes()
```

Maybe its better to think about the case of Edge Plots. For visualizing geometry, we would still have:
```Python
uxds.uxgrid.plot.edges()
```

But how about if we have a data variable on the edge (i.e. gradients). It would simiarly be called
```Python
uxds.['grad'].plot.edges()
```

They are both "edges", but one has a data variable and the other doesn't.

Reply:
I agree with that "Having `UxDataArray.plot.values()` return a node plot with face-centered data may be misleading". I also agree with the default `UxDataArray.plot()` method covering a similar need to what I suggested with `UxDataArray.plot.values()`, so let's stick to `UxDataArray.plot()` only. 

However, I don't agree with the idea of  "Plot Nodes with Data" via `uxds['d_var'].plot.nodes()`; I think that'd also be misleading in the user's end (See the next paragraph for an example case). I believe we should consistently use the names "nodes" and "edges" for plotting functions (no what what class, Grid, UxDataArray, or UxDataset, they are called from) for the grid geometry information only (i.e. without any information about the data variable). In my suggestion, `uxds.uxgrid.plot.nodes()` and `uxds['d_var'].plot.nodes()` are and should be exactly equal (Just because not accessing `.uxgrid` accessor, I don't think the end user would expect the difference you propose; I think they'd still expect to explore the grid geometry). 

Having a face-centered data variable and calling `uxds['d_var'].plot.nodes()` would give a `ValueError` in the current implementation, which would be confusing for the end-user since they'd expect to explore the geometry of the Grid that is associated to their UxDataArray.

cc'ing @rajeeja if he has any thoughts.

Reply:
Also, re "But how about if we have a data variable on the edge (i.e. gradients)", the default ` uxda.plot()` function should work very well for any of these possibilities, i.e. face-centered, node-centered, edge-centered, and is named very intuitively. One would most likely expect `UxDataArray.plot()` to plot the actual data values.

Reply:
> Having a face-centered data variable and calling uxds['d_var'].plot.nodes() would give a ValueError in the current implementation, which would be confusing for the end-user since they'd expect to explore the geometry of the Grid that is associated to their UxDataArray.

Is this not desired? The way we've written most (if not all) of our functionality is rooted in this separation between `Grid` and `UxDataArray` / `UxDataset` functionality. 

For someone exploring the geometry of the Grid is expected to either have a standalone `Grid` instance or use the `uxda.uxgrid` / `uxds.uxgrid` accessor. This is where our `GridPlotAccessor` comes in, which enables our strictly "geometric" plotting routines through `Grid.plot()` and other methods. 

When interacting with data variables, everything is handled through the `UxDataset` and `UxDataArray`, with most of the calls to the underlying `uxgrid` accessor abstracted away. 

Given this, while our "node" in both the `Grid` and `UxDataArray` case both share the same spatial information, the `uxda.plot.node` method enables the coloring of each node, if applicable (corner node data).

Reply:
> Having a face-centered data variable and calling uxds['d_var'].plot.nodes() would give a ValueError in the current implementation, which would be confusing for the end-user since they'd expect to explore the geometry of the Grid that is associated to their UxDataArray.

The more I think about this, the more it makes sense to split up `plot.nodes()` into `plot.corner_nodes()` and `plot.centroids` or `plot.face_centers()`.

This explicit naming is much more clear, and it would still enable us to plot face-centered data variables respective to the face center coordinates.

I think geometrically it makes sense too, respective to our `Grid`.

Reply:
> Also, re "But how about if we have a data variable on the edge (i.e. gradients)", the default  uxda.plot() function should work very well for any of these possibilities, i.e. face-centered, node-centered, edge-centered, and is named very intuitively. One would most likely expect UxDataArray.plot() to plot the actual data values.

Our default `plot` method will attempt to chose the most appropriate plotting method based on the mapping of the data (additionally the size of the data will dictate whether we default to vector or rasterized approaches)

Reply:
To summarize my thoughts, one should never expect to strictly visualize the "geometry" of the grid if they access they use the `.plot()` method of a `UxDataset` or `UxDataArray`. These will _always_ plot some data variable residing on our unstructured grid (or through an alternative visualuzation method such as computing the trimesh).

Reply:
> To summarize my thoughts, one should never expect to strictly visualize the "geometry" of the grid if they access they use the .plot() method of a UxDataset or UxDataArray. These will always plot some data variable residing on our unstructured grid (or through an approximate visualization method such as computing the trimesh).

Or we should avoid using the same naming for different functionality in the plotting context (or in general). This is not only due to expectations but also due to the high possibility of one's overlooking if there is `.ugrid` or not in their code (e.g. `uxds[`var`].uxgrid.plot.nodes()` looks quite similar to `uxds[`var`].plot.nodes()`, especially in a long workflow, and can easily be overlooked). To make the distinction much clearer, the plotting function with data variable information could be named differently, something like:

`UxDataArray.plot.node_values()`

Reply:
> The more I think about this, the more it makes sense to split up `plot.nodes()` into `plot.corner_nodes()` and `plot.centroids` or `plot.face_centers()`.
> 
> This explicit naming is much more clear, and it would still enable us to plot face-centered data variables respective to the face center coordinates.
> 
> I think geometrically it makes sense too, respective to our `Grid`.

I don't agree with this since "nodes" are already the "corner nodes". If we call them "corner nodes", then one would get confused about if there are other types of nodes in the grid geometry, but there is not. Having only "nodes", "edges", and "face_centers" is sufficient IMO.

Reply:
> > Having a face-centered data variable and calling uxds['d_var'].plot.nodes() would give a ValueError in the current implementation, which would be confusing for the end-user since they'd expect to explore the geometry of the Grid that is associated to their UxDataArray.
> 
> Is this not desired? The way we've written most (if not all) of our functionality is rooted in this separation between `Grid` and `UxDataArray` / `UxDataset` functionality.
> 
> For someone exploring the geometry of the Grid is expected to either have a standalone `Grid` instance or use the `uxda.uxgrid` / `uxds.uxgrid` accessor. This is where our `GridPlotAccessor` comes in, which enables our strictly "geometric" plotting routines through `Grid.plot()` and other methods.
> 
> When interacting with data variables, everything is handled through the `UxDataset` and `UxDataArray`, with most of the calls to the underlying `uxgrid` accessor abstracted away.
> 
> Given this, while our "node" in both the `Grid` and `UxDataArray` case both share the same spatial information, the `uxda.plot.node` method enables the coloring of each node, if applicable (corner node data).

This makes much sense to me, and I agree with how we mostly differentiate the ` uxda` and `uxda.uxgrid` from each other, but as I explained in the second prior message, using the exact same name (i.e. `.nodes()`) would still confuse.  I'd highly recommend to use a different name than the particular geometric element's name to make the distinction 100% clear.

Reply:
> > Also, re "But how about if we have a data variable on the edge (i.e. gradients)", the default  uxda.plot() function should work very well for any of these possibilities, i.e. face-centered, node-centered, edge-centered, and is named very intuitively. One would most likely expect UxDataArray.plot() to plot the actual data values.
> 
> Our default `plot` method will attempt to chose the most appropriate plotting method based on the mapping of the data (additionally the size of the data will dictate whether we default to vector or rasterized approaches)

Yes, and I feel like this is the perfect way of working. However, it'd be fine to have additional functions like `uxda.plot.node_values()` `uxda.plot.edge_values()`, and `uxda.plot.face_values()` (or alternative names that are not the same as Grid's geometric element plotting fucns' names), that would cover parts of this default `plot()` or throw `ValueError` if not applicable. 

Would it make sense?

Reply:
> > The more I think about this, the more it makes sense to split up `plot.nodes()` into `plot.corner_nodes()` and `plot.centroids` or `plot.face_centers()`.
> > This explicit naming is much more clear, and it would still enable us to plot face-centered data variables respective to the face center coordinates.
> > I think geometrically it makes sense too, respective to our `Grid`.
> 
> I don't agree with this since "nodes" are already the "corner nodes". If we call them "corner nodes", then one would get confused about if there are other types of nodes in the grid geometry, but there is not. Having only "nodes", "edges", and "face_centers" is sufficient IMO.

Maybe we should stray away from calling this method "nodes", since like you mentioned we reffer to "nodes" as our "corner nodes".

We have three types of elements that could be interpreted as "nodes":
* `node_lat/lon`: Corner Node Coordinates
* `face_lat/lon`: Face Center Coordinates
* `edge_lat/lon`: Edge Center Coordinates

Would it make more sense to call it `Grid.plot.points(element="nodes")` and specify which element ("nodes", "faces", "edges"). This also lines up with our "Point Rasterization" instead of calling it "Node Rasterization".

Reply:
We could also keep `Grid.plot.nodes` as an alias of `Grid.plot.points(element="nodes")`, while re-writing `UxDataArray.plot.nodes` to `UxDataArray.plot.points` with a similar naming convention.

Reply:
> > Our default `plot` method will attempt to chose the most appropriate plotting method based on the mapping of the data (additionally the size of the data will dictate whether we default to vector or rasterized approaches)
> 
> Yes, and I feel like this is the perfect way of working. However, it'd be fine to have additional functions like `uxda.plot.node_values()` `uxda.plot.edge_values()`, and `uxda.plot.face_values()` (or alternative names that are not the same as Grid's geometric element plotting fucns' names), that would cover parts of this default `plot()` or throw `ValueError` if not applicable.
> 
> Does it make sense?

This does make sense, but personally I'm not a big fan of extending the names with an `_` I'd like to hear what @rajeeja, @rljacob , or anyone else thinks about the `uxgrid` accessor not being enoguh to difference between "grid-only" and "data variable" plotting.

Reply:
> Maybe we should stray away from calling this method "nodes", since like you mentioned we reffer to "nodes" as our "corner nodes".
> 
> We have three types of elements that could be interpreted as "nodes":
> 
> * `node_lat/lon`: Corner Node Coordinates
> * `face_lat/lon`: Face Center Coordinates
> * `edge_lat/lon`: Edge Center Coordinates
> 
> Would it make more sense to call it `Grid.plot.points(element="nodes")` and specify which element ("nodes", "faces", "edges"). This also lines up with our "Point Rasterization" instead of calling it "Node Rasterization".

To me (and the UGRID conventions AFAIK), the only one that should actually be interpreted as "nodes" is the first one out of these three. The others should never be interpreted as "nodes", maybe instead "face_centers (or face_coordinates)" and "edge_centers (or edge_coordinates)". That said I would not find `Grid.plot.points(element="nodes")` to be better.

Could you please elaborate on why you'd interprete all these three as "nodes"?

Reply:
Additionally, my general understanding is that "nodes" or "vertices" (in different models etc.) are always interpreted for the "corner nodes" and never for anything related to face or edges. Similarly, "face" or "cell" is always used for what we know as face in our Grid. MPAS has "vertices"  of Dual esh corresponding to "face_centrers" of Primal mesh, and vice versa, but that's a special case.

Reply:
> To me (and the UGRID conventions AFAIK), the only one that should actually be interpreted as "nodes" is the first one out of these three. The others should never be interpreted as "nodes", maybe instead "face_centers (or face_coordinates)" and "edge_centers (or edge_coordinates)". That said I would not find Grid.plot.points(element="nodes") to be better.
> 
> Could you please elaborate on why you'd interprete all these three as "nodes"?

I consider a "node", "vertex", or "point" in our context as a tuple of `(lon, lat)`. Each of our `node_lat/lon`, `face_lat/lon` and `edge_lat/lon` are in this form. Below are some snippets from the MPAS specification.

![image](https://github.com/UXARRAY/uxarray/assets/67855069/ecd13105-e1d3-4004-b4f2-5b29821252af)

![image](https://github.com/UXARRAY/uxarray/assets/67855069/8b59dbed-2107-4962-b54a-e8b28fff67b0)

![image](https://github.com/UXARRAY/uxarray/assets/67855069/5d78a3f2-ec40-4be6-96d8-10fb707be497)

Reply:
One scenario where we may want to plot `node_lat/lon` alongside `face_lat/lon` is when investigating the computed centroid location.

Or similarly, plotting the `edge_lat/lon` with a mesh/edge plot to see if they centroids lie on each edge.

Reply:
For the DataArray case, one may want to plot an `edge` data variable as a note, instead of shading the whole edge, when plotting the gradient for example.

Reply:
> > To me (and the UGRID conventions AFAIK), the only one that should actually be interpreted as "nodes" is the first one out of these three. The others should never be interpreted as "nodes", maybe instead "face_centers (or face_coordinates)" and "edge_centers (or edge_coordinates)". That said I would not find Grid.plot.points(element="nodes") to be better.
> > Could you please elaborate on why you'd interprete all these three as "nodes"?
> 
> I consider a "node", "vertex", or "point" in our context as a tuple of `(lon, lat)`. Each of our `node_lat/lon`, `face_lat/lon` and `edge_lat/lon` are in this form. Below are some snippets from the MPAS specification.
> 
> ![image](https://user-images.githubusercontent.com/67855069/282960807-ecd13105-e1d3-4004-b4f2-5b29821252af.png)
> 
> ![image](https://user-images.githubusercontent.com/67855069/282960822-8b59dbed-2107-4962-b54a-e8b28fff67b0.png)
> 
> ![image](https://user-images.githubusercontent.com/67855069/282960837-5d78a3f2-ec40-4be6-96d8-10fb707be497.png)

I see. I'd agree with you (i.e. naming tuples of lot/lan as "nodes" regardless of them being faces or edges as well) if this was a general-purpose geometry application, but since UXarray is specialized to unstructured grids and in geoscientific unstructured grid applications "node" is used for the "corner of a face", I don't agree with this interpretation (i.e. we should avoid confusing "nodes" with such tuples). 

I think even the MPAS snippets you provided support the fact that "vertices/nodes" are used only for the "corner of cells/faces", not for any face-related or edge-related thing (despite MPAS's special case I mentioned).

For instance, the text of `latCell` and `lonCell` only refers to "cell centers", that of `latEdge` and `lonEdge` only mentions to "edge locations"; however, `latVertex` and `lonVertex` mentions "cell vertices".

When it comes to tuples of such lat/lons, amybe one should call them "node_coords", "edge_coords", "face_coords", not "nodes" or "vertices". 

I am happy to discuss this in a video chat, 1:1 or in a broader group.

Reply:
Calling them "cords" makes much more sense to me.

Reply:
> One scenario where we may want to plot the `face_lat/lon` is when we want to investigate the centroid computation. We would plot these as "nodes" along with a `edge` or `polygon` plot
> 
> ```python
> # composition with Holoviews 
> uxgrid.mesh() * uxgrid.points(element='face')
> ```
> 
> Or similarly, plotting the `edge_lat/lon` with a mesh/edge plot to see if they centroids lie on each edge.
> 
> ```python
> uxgrid.mesh() * uxgrid.points(element='edge')
> ```

I'd foresee these as:

`uxgrid.mesh() * uxgrid.face_centers()` and 
`uxgrid.mesh() * uxgrid.edge_centers()` or `uxgrid.mesh() * uxgrid.edge_locations()`

Reply:
> Calling them "coords" makes much more sense to me.

Yeah, or "nodes", "face_centers", and "edge_locations" always sounded sufficient to me.